src/tiled/moveselectiontool.cpp

```diff
@@ -0,0 +1,184 @@
+                       parent)
+    , mDragging(false)
+    , mMouseDown(false)
+{
+}
+
+MoveSelectionTool::~MoveSelectionTool()
+{
+}
+
+void MoveSelectionTool::activate(MapScene *scene)
+{
+    AbstractTileTool::activate(scene);
+    cut();
+}
+
+void MoveSelectionTool::deactivate(MapScene *scene)
+{
+    AbstractTileTool::deactivate(scene);
+    paste();
+}
+
+void MoveSelectionTool::tilePositionChanged(const QPoint &pos)
+{
+    if (mDragging) {
+        QPoint offset = pos - mLastUpdate;
+        QRegion selectedArea = mapDocument()->selectedArea();
+        selectedArea.translate(offset);
+
+        mapDocument()->undoStack()->push(new ChangeSelectedArea(mapDocument(), selectedArea));
+
+        if (mPreviewLayer) {
+            mPreviewLayer->setX(mPreviewLayer->x()+offset.x());
+            mPreviewLayer->setY(mPreviewLayer->y()+offset.y());
+            brushItem()->setTileLayer(mPreviewLayer);
+            brushItem()->setTileRegion(mapDocument()->selectedArea());
+        }
+
+        mLastUpdate = pos;
+    }
+}
+
+void MoveSelectionTool::mouseMoved(const QPointF &pos, Qt::KeyboardModifiers modifiers)
+{
+    if (mMouseDown && !mDragging) {
+        QPoint screenPos = QCursor::pos();
+        const int dragDistance = (mMouseScreenStart - screenPos).manhattanLength();
+
+        if (dragDistance >= QApplication::startDragDistance() / 2) {
+            mDragging = true;
+            tilePositionChanged(tilePosition());
+        }
+    }
+
+    AbstractTileTool::mouseMoved(pos, modifiers);
+
+    refreshCursor();
+}
+
+void MoveSelectionTool::mousePressed(QGraphicsSceneMouseEvent *event)
+{
+    if (event->button() == Qt::LeftButton && mapDocument()->selectedArea().contains(tilePosition())) {
+        mMouseScreenStart = event->screenPos();
+        mDragStart = tilePosition();
+        mMouseDown = true;
+        mLastUpdate = tilePosition();
+    }
+}
+
+void MoveSelectionTool::mouseReleased(QGraphicsSceneMouseEvent *)
+{
+    if (mDragging) {
+        mDragging = false;
+        brushItem()->setTileRegion(QRegion());
+    }
+
+    mMouseDown = false;
+    refreshCursor();
+}
+
+void MoveSelectionTool::languageChanged()
+{
+    setName(tr("Move Selection"));
+    setShortcut(QKeySequence(tr("G")));
+}
+
+void MoveSelectionTool::refreshCursor()
+{
+    Qt::CursorShape cursorShape = Qt::ArrowCursor;
+
+    if (mDragging)
+        cursorShape = Qt::SizeAllCursor;
+
+    if (cursor().shape() != cursorShape)
+        setCursor(cursorShape);
+}
+
+void MoveSelectionTool::cut()
+{
+    Layer *currentLayer = mapDocument()->currentLayer();
+
+    TileLayer *tileLayer = dynamic_cast<TileLayer*>(currentLayer);
+    const QRegion &selectedArea = mapDocument()->selectedArea();
+
+    TileLayer *brushLayer = tileLayer->copy(tileLayer->bounds());
+    mPreviewLayer = SharedTileLayer(brushLayer);
+
+    brushItem()->setTileLayer(mPreviewLayer);
+    brushItem()->setTileRegion(mapDocument()->selectedArea());
```

Comment:
Please use the version of `setTileLayer` that also takes a region, otherwise `setTileLayer` will first compute the region from the tile layer, which is a waste of time:
```c++
brushItem()->setTileLayer(mPreviewLayer, selectedArea);
```